### PR TITLE
IGNITE-14871 Add cluster initialization commands to the Ignite CLI

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-14871

### CODE REVIEW

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/InitClusterRequest.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+package org.apache.ignite.cli.builtins.cluster;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.List;
+
+/**
+ * A request to the Cluster Init REST API.
+ */
+public class InitClusterRequest {
+    @SuppressWarnings("unused")
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/InitClusterRequest.java:27]: I think you don't need to suppress anything here, you can set up IDEA to ignore fields annotated with `@JsonProperty`

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    @Inject
+    public ClusterApiClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * Sends 'cluster init' command to the specified node.
+     *
+     * @param nodeEndpoint host:port on which REST API is listening
+     * @param metastorageNodeIds consistent IDs of the nodes that will form metastorage RAFT group
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:57]:
```suggestion
     * @param metastorageNodeIds consistent IDs of the nodes that will host the Meta Storage Raft group
```

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     * @param cmgNodeIds         consistent IDs of the nodes that will form Cluster Management RAFT Group; if empty,
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:58]:
```suggestion
     * @param cmgNodeIds         consistent IDs of the nodes that will host the Cluster Management Raft Group; if empty,
```

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+     *                           then nodes specified in metastorageNodeIds will form Cluster Management RAFT Group as well
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:59]: I would suggest to rephrase this part:
```
If omitted, {@code metastorageNodeIds} will be used to host the CMG as well.

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    private IgniteCliException error(String message, HttpResponse<String> httpResponse) {
+        String errorPayload;
+        try {
+            errorPayload = prettifyJson(httpResponse);
+        } catch (JsonProcessingException e) {
+            // not a valid JSON probably
+            errorPayload = httpResponse.body();
+        }
+
+        return new IgniteCliException(message + "\n\n" + errorPayload);
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:104]: Does "\n\n" work on Windows?

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Sends 'cluster init' command to the specified node.
+     *
+     * @param nodeEndpoint host:port on which REST API is listening
+     * @param metastorageNodeIds consistent IDs of the nodes that will form metastorage RAFT group
+     * @param cmgNodeIds         consistent IDs of the nodes that will form Cluster Management RAFT Group; if empty,
+     *                           then nodes specified in metastorageNodeIds will form Cluster Management RAFT Group as well
+     * @param out                {@link PrintWriter} to which to report about the command outcome
+     */
+    public void init(String nodeEndpoint, List<String> metastorageNodeIds, List<String> cmgNodeIds, PrintWriter out) {
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:62]: Do you validate the `nodeEndpoint` format anywhere?

File: modules/cli/src/main/java/org/apache/ignite/cli/spec/ClusterCommandSpec.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                description = "Address of the REST endpoint of the receiving node in host:port format")
+        private String nodeEndpoint;
+
+        /**
+         * List of names of the nodes (each represented by a separate command line argument) that will host the Metastorage Raft group.
+         * If the "--cmg-nodes" parameter is omitted, the same nodes will also host the Cluster Management Raft group.
+         */
+        @Option(names = "--meta-storage-node", required = true, description = {
+                "Name of the node (repeat like '--meta-storage-node node1 --meta-storage-node node2' to specify more than one node)",
+                "that will host the Metastorage Raft group.",
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/spec/ClusterCommandSpec.java:60]:
```suggestion
                "that will host the Meta Storage Raft group.",
```

File: modules/cli/src/main/java/org/apache/ignite/cli/spec/ClusterCommandSpec.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        @Inject
+        private ClusterApiClient clusterApiClient;
+
+        /**
+         * Address of the REST endpoint of the receiving node in host:port format.
+         */
+        @Option(names = "--node-endpoint", required = true,
+                description = "Address of the REST endpoint of the receiving node in host:port format")
+        private String nodeEndpoint;
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/spec/ClusterCommandSpec.java:52]: Should you use `NodeEndpointOptions` here?

File: modules/cli/src/main/java/org/apache/ignite/cli/spec/NodeEndpointOptions.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+     * Parses REST endpoint host and port from string.
+     *
+     * @return 2-elements array [host, port].
+     */
+    private String[] parse() {
+        var hostPort = endpoint.split(":");
+
+        if (hostPort.length != 2) {
+            throw new IgniteCliException("Incorrect host:port pair provided "
+                    + "(example of valid value 'localhost:10300')");
```
Comment (sashapolo) [inline modules/cli/src/main/java/org/apache/ignite/cli/spec/NodeEndpointOptions.java:75]: I would suggest to also add the provided value to the message

File: modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java [hunk 10]
```diff
@@ -581,17 +619,200 @@ void setJson() throws IOException, InterruptedException {
...
+            assertThatStderrIsEmpty();
+        }
+    }
+
+    /**
+     * Tests "cluster" command.
+     */
+    @Nested
+    @DisplayName("cluster")
+    class Cluster {
```
Comment (sashapolo) [inline modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: Is it possible to extract this class to an upper level?

Comment (sashapolo) [inline modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: IDEA complains that this class is not called something like `ClusterTest`

File: modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java [hunk 11]
```diff
@@ -603,10 +824,10 @@ void setJson() throws IOException, InterruptedException {
      * @param actual Actual result.
      * @see Object#equals(Object)
      */
-    private static void assertEquals(String exp, String actual) {
-        Assertions.assertEquals(
-                exp.lines().collect(Collectors.toList()),
-                actual.lines().collect(Collectors.toList())
+    private static void assertOutputEqual(String exp, String actual) {
+        assertEquals(
+                exp.lines().collect(toList()),
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:829]: It worked like this before me, so I just did not want to change the logic without a good reason. Maybe the guys wanted to remove newlines.

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/InitClusterRequest.java (hunk 1 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/InitClusterRequest.java:27]: I though similarly, but PMD has its own opinion :) It just fails the build at TC if it sees an unused field.

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java (hunk 5 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java]: No it doesn't, good catch. Fixed this.

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java (hunk 6 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:62]: No, the user will see an exception with the full URL containing any strange thing they will specify here. Do you believe we need to validatee it manually ourselves?

File: modules/cli/src/main/java/org/apache/ignite/cli/spec/ClusterCommandSpec.java (hunk 8 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/spec/ClusterCommandSpec.java:52]: `NodeEndpointOptions` allows the endpoint to be omitted. In the case of the init command, it seems super-important to make sure that the user knows exactly what endpoint they use, so I made it mandatory, hence `NodeEndpointOptions` cannot be used.

File: modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java (hunk 10 above)
Comment (rpuch) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: It uses some infrastructure defined in the containing class, that would probably require to transform all the internal test classes defined here for consistency. Probably a separaate JIRA issue for refactoring tests seems a good option, WDYT?

Comment (rpuch) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: Yes, it follows the pattern given by its neighbors. I'm not sure whether it is a good idea to change it in this task (see the previous comment, maybe a separate JIRA ticket is an option).

File: modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java [hunk 12]
```diff
@@ -581,17 +619,200 @@ void setJson() throws IOException, InterruptedException {
...
+        @DisplayName("init --node-endpoint=127.0.0.1:17300 --meta-storage-node node1ConsistentId --meta-storage-node node2ConsistentId "
+                + "--cmg-node node2ConsistentId --cmg-node node3ConsistentId")
+        void initSuccess() throws Exception {
+            when(response.statusCode()).thenReturn(HttpURLConnection.HTTP_OK);
+            when(httpClient.<String>send(any(), any())).thenReturn(response);
+
+            var expSentContent = "{\"metastorageNodes\":[\"node1ConsistentId\",\"node2ConsistentId\"],"
+                    + "\"cmgNodes\":[\"node2ConsistentId\",\"node3ConsistentId\"]}";
+
+            var cmd = cmd(ctx);
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java]: Fixed

File: modules/cli/src/main/java/org/apache/ignite/cli/spec/NodeEndpointOptions.java (hunk 9 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/spec/NodeEndpointOptions.java:75]: Added it to the exception message

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java (hunk 4 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java]: Applied your suggestion

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java (hunk 6 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:62]: Here is how it looks now:

```
java.lang.IllegalArgumentException: Illegal character in authority at index 7: http://some shit/management/v1/cluster/init/
```

File: modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java (hunk 10 above)
Comment (sashapolo) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: a ticket seems fine

Comment (sashapolo) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: ok!

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java (hunk 6 above)
Comment (sashapolo) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/cluster/ClusterApiClient.java:62]: How do other input errors look like? If they look similarly, then I'm fine with the current approach

File: modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java (hunk 10 above)
Comment (rpuch) [reply modules/cli/src/test/java/org/apache/ignite/cli/IgniteCliInterfaceTest.java:644]: I filed a ticket https://issues.apache.org/jira/browse/IGNITE-16520

**Pull request merged**